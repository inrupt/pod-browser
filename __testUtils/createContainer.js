/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  createSolidDataset,
  createThing,
  addDatetime,
  addDecimal,
  addInteger,
  setThing,
} from "@inrupt/solid-client";
import { rdf, ldp } from "rdf-namespaces";

export const TIMESTAMP = new Date(Date.UTC(2020, 5, 2, 15, 59, 21));

export default function createContainer(
  url = "https://user.dev.inrupt.net/",
  type = ldp.BasicContainer
) {
  const thing = createThing({ url });
  const publicContainer = [
    (t) => addUrl(t, rdf.type, type),
    (t) => addUrl(t, rdf.type, ldp.Container),
    (t) => addDatetime(t, "http://purl.org/dc/terms/modified", TIMESTAMP),
    (t) =>
      addDecimal(t, "http://www.w3.org/ns/posix/stat#mtime", 1591131561.195),
    (t) => addInteger(t, "http://www.w3.org/ns/posix/stat#size", 4096),
    (t) => addUrl(t, ldp.contains, "https://user.dev.inrupt.net/public/games/"),
  ].reduce((acc, fn) => fn(acc), thing);

  return setThing(createSolidDataset(), publicContainer);
}
