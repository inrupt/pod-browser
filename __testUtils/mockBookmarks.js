/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addDatetime,
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { dct, rdf } from "rdf-namespaces";
import {
  RECALLS_PROPERTY_IRI,
  BOOKMARK_TYPE_IRI,
} from "../src/solidClientHelpers/bookmarks";

const bookmarksUrl = "http://example.com/bookmarks/index.ttl";
const bookmark1Title = "Cat gifs";
const bookmark1Recalls = "https://example.com/catgifs";
const bookmark2Title = "Dog videos";
const bookmark2Recalls = "https://example.com/dogvideos";

export default function mockBookmarks() {
  let bookmarksDataset = mockSolidDatasetFrom(bookmarksUrl);
  const bookmark1 = mockThingFrom(`${bookmarksUrl}#1234`);
  const bookmark1WithTitle = addStringNoLocale(
    bookmark1,
    dct.title,
    bookmark1Title
  );
  const bookmark1WithRecalls = addUrl(
    bookmark1WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark1Recalls
  );
  const bookmark1WithType = addUrl(
    bookmark1WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark1WithCreated = addDatetime(
    bookmark1WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 28, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark1WithCreated);
  const bookmark2 = mockThingFrom(`${bookmarksUrl}#4567`);
  const bookmark2WithTitle = addStringNoLocale(
    bookmark2,
    dct.title,
    bookmark2Title
  );
  const bookmark2WithRecalls = addUrl(
    bookmark2WithTitle,
    RECALLS_PROPERTY_IRI,
    bookmark2Recalls
  );
  const bookmark2WithType = addUrl(
    bookmark2WithRecalls,
    rdf.type,
    BOOKMARK_TYPE_IRI
  );
  const bookmark2WithCreated = addDatetime(
    bookmark2WithType,
    RECALLS_PROPERTY_IRI,
    new Date(Date.UTC(2020, 9, 29, 3, 24, 0))
  );
  bookmarksDataset = setThing(bookmarksDataset, bookmark2WithCreated);
  return {
    dataset: bookmarksDataset,
    iri: bookmarksUrl,
  };
}
