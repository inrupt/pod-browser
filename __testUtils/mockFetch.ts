/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import "whatwg-fetch";
import Mock = jest.Mock;

type fetchType = typeof window.fetch;

export default function mockFetch(
  responses: Record<string, Response | Record<string, Response>>
): Mock<fetchType> {
  return jest
    .fn<any, Parameters<fetchType>>()
    .mockImplementation((url, init) => {
      const response = responses[url.toString()];
      const responseObject = response as Record<string, Response>;
      const method = init?.method || "GET";
      if (typeof responseObject[method] !== "undefined") {
        return Promise.resolve(responseObject[method]);
      }
      if (response) {
        return Promise.resolve(response);
      }
      throw new Error(`URL (${url}) not mocked properly`);
    });
}

export function mockSolidClientRequest(
  body: string,
  method = "GET",
  headers: Record<string, string> = {}
): Request {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status: number, body = ""): Response {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
