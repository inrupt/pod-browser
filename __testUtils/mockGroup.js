/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  getStringNoLocale,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/group";
import { vcardExtras } from "../src/models/addressBook";

function mockGroupThing(name, groupThingUrl) {
  return chain(
    mockThingFrom(groupThingUrl),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, vcard.fn, name)
  );
}

function mockIndexThing(addressBook, groupThingUrl) {
  return setUrl(addressBook.thing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
  );
}

// export function addGroupsToAddressBook(addressBook, groupsToAdd) {
//   const { groups, ...rest } = addressBook;
//   return {
//     ...rest,
//     groups: {
//       iri: addressBook.groups.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...groupsToAdd.map(({ dataset, iri }) => (d) => {
//           const group = getThing(dataset, iri);
//           const name = getStringNoLocale(group, vcard.fn);
//           return setThing(d, mockGroupThing(name, iri));
//         }),
//         ...groupsToAdd.map((group) => (d) =>
//           setThing(d, mockIndexThing(addressBook, group.iri))
//         )
//       ),
//     },
//   };
// }

export default function mockGroup(
  addressBook,
  name,
  { url, id, members } = {}
) {
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = chain(
    mockGroupThing(name, groupThingUrl),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    )
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
