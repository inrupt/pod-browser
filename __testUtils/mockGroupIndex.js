/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { mockSolidDatasetFrom } from "@inrupt/solid-client";
import { getContactIndexDefaultUrl } from "../src/models/contact/collection";
import { addGroupToMockedIndexDataset } from "./mockGroupContact";
import { getGroupName, getGroupUrl } from "../src/models/group";
import { chain } from "../src/solidClientHelpers/utils";
import { GROUP_CONTACT } from "../src/models/contact/group";

export default function mockGroupIndex(addressBook, groups) {
  const contactIndexUrl = getContactIndexDefaultUrl(
    addressBook.containerUrl,
    GROUP_CONTACT
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(contactIndexUrl),
      ...groups.map((g) => (d) =>
        addGroupToMockedIndexDataset(
          d,
          addressBook,
          getGroupName(g),
          getGroupUrl(g)
        )
      )
    ),
    type: GROUP_CONTACT,
  };
}
