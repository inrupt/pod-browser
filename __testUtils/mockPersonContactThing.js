/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, foaf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/models/addressBook";

export const webIdUrl = "http://example.com/alice#me";

export function addMockedPersonThingsToIndexDataset(
  dataset,
  addressBook,
  name,
  personThingUrl
) {
  return chain(dataset, (d) =>
    setThing(
      d,
      chain(
        mockThingFrom(personThingUrl),
        (t) => setUrl(t, rdf.type, vcard.Individual),
        (t) => setStringNoLocale(t, vcard.fn, name),
        (t) => setUrl(t, vcardExtras("inAddressBook"), asUrl(addressBook.thing))
      )
    )
  );
}

// export function addContactsToAddressBook(addressBook, contactsToAdd) {
//   // TODO: Should probably be moved to a separate mockContact.js
//   const { people, ...rest } = addressBook;
//   return {
//     ...rest,
//     people: {
//       iri: addressBook.people.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...contactsToAdd.map((contactThing) => (d) =>
//           setThing(
//             d,
//             chain(contactThing, (t) =>
//               setUrl(t, vcardExtras("inAddressBook"), addressBook.index.iri)
//             )
//           )
//         )
//       ),
//     },
//   };
// }

export default function mockPersonContactThing(url = webIdUrl) {
  return chain(
    mockThingFrom(url),
    (t) => addUrl(t, rdf.type, vcard.Individual),
    (t) => addUrl(t, foaf.openid, webIdUrl)
  );
}
