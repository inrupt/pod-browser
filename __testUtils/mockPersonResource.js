/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { addStringNoLocale, addUrl, mockThingFrom } from "@inrupt/solid-client";
import { vcard, foaf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";

export const aliceWebIdUrl = "http://example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockPersonDatasetAlice() {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto)
  );
}

export function mockProfileAlice() {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice());
}

export const bobWebIdUrl = "http://example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";

export function mockPersonDatasetBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick)
  );
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
