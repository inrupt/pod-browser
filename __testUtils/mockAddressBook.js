/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { acl, dc, rdf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { aliceWebIdUrl } from "./mockPersonResource";
import { vcardExtras } from "../src/addressBook";
import { getAddressBookIndexDefaultUrl } from "../src/models/addressBook";
import { joinPath } from "../src/stringHelpers";

export const defaultContainerUrl = "https://user.example.com/contacts";
export const defaultAddressBookDatasetUrl = joinPath(
  defaultContainerUrl,
  "index.ttl"
);
export const defaultAddressBookThingUrl = `${defaultAddressBookDatasetUrl}#this`;

export function mockAddressBookThing(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const owner = options.owner || aliceWebIdUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  const mainIndexUrl = `${indexUrl}#this`;
  return chain(
    mockThingFrom(mainIndexUrl),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, "Contacts")
  );
}

export function mockAddressBookDataset(
  options = {},
  thing = mockAddressBookThing(options)
) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const indexUrl =
    options.indexUrl || getAddressBookIndexDefaultUrl(containerUrl);
  return chain(mockSolidDatasetFrom(indexUrl), (d) => setThing(d, thing));
}

export default function mockAddressBook(options = {}) {
  const containerUrl = options.containerUrl || defaultContainerUrl;
  const thing = mockAddressBookThing(options);
  return {
    containerUrl,
    thing,
    dataset: mockAddressBookDataset(options, thing),
  };
}

export function mockEmptyAddressBook(containerUrl, owner, title = "Contacts") {
  const thing = chain(
    mockAddressBookThing({ containerUrl, owner }),
    (t) => addUrl(t, rdf.type, vcardExtras("AddressBook")),
    (t) => addUrl(t, acl.owner, owner),
    (t) => addStringNoLocale(t, dc.title, title)
  );
  return {
    containerUrl,
    dataset: setThing(createSolidDataset(), thing),
    thing,
  };
}
