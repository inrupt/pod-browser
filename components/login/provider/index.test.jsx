/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { mount } from "enzyme";

import {
  clearLocalstorage,
  generateRedirectUrl,
} from "../../../src/windowHelpers";

import SessionContext from "../../../src/contexts/sessionContext";
import { mountToJson, WithTheme } from "../../../__testUtils/mountWithTheme";

import ProviderLogin, * as ProviderFunctions from "./index";

jest.mock("../../../src/windowHelpers");

describe("ProviderLogin form", () => {
  test("Renders a webid login form, with button bound to login", () => {
    const tree = mountToJson(<ProviderLogin />);

    expect(tree).toMatchSnapshot();
  });
});

describe("loginWithProvider", () => {
  test("clicking login calls loginWithProvider", () => {
    const session = {
      login: jest.fn(),
    };

    const tree = mount(
      <SessionContext.Provider value={{ session }}>
        <WithTheme>
          <ProviderLogin />
        </WithTheme>
      </SessionContext.Provider>
    );

    tree
      .find("button[children='Log In']")
      .simulate("click", { preventDefault: () => {} });

    expect(session.login).toHaveBeenCalled();
  });

  test("Calls login", async () => {
    const redirect = "foo";
    const providerIri = "https://test.url";
    const session = {
      login: jest.fn(),
    };

    generateRedirectUrl.mockReturnValue("foo");

    await ProviderFunctions.loginWithProvider(providerIri, session);

    expect(clearLocalstorage).toHaveBeenCalled();
    expect(session.login).toHaveBeenCalledWith({
      oidcIssuer: providerIri,
      redirectUrl: redirect,
    });
  });

  test("Bubbles errors", async () => {
    const error = "Failure";
    const providerIri = "https://test.url";
    const session = {
      login: jest.fn().mockRejectedValue(error),
    };

    await expect(
      ProviderFunctions.loginWithProvider(providerIri, session)
    ).rejects.toMatch(error);
  });
});
