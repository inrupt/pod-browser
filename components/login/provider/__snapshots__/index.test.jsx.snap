// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProviderLogin form Renders a webid login form, with button bound to login 1`] = `
<DocumentFragment>
  <form>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <div
        class="PodBrowser-root PodBrowser-root"
      >
        <h3
          data-testid="login-title"
        >
          Log In
        </h3>
        <div
          aria-expanded="false"
          class="PodBrowser-root PodBrowser-hasClearIcon"
          role="combobox"
        >
          <div
            class="PodBrowser-root PodBrowser-root PodBrowser-marginNormal PodBrowser-fullWidth"
          >
            <label
              class="PodBrowser-root PodBrowser-root PodBrowser-formControl PodBrowser-animated PodBrowser-outlined PodBrowser-required PodBrowser-required"
              data-shrink="false"
              for="provider-select"
              id="provider-select-label"
            >
              Select ID provider
              <span
                aria-hidden="true"
                class="PodBrowser-asterisk PodBrowser-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="PodBrowser-root PodBrowser-root PodBrowser-inputRoot PodBrowser-fullWidth PodBrowser-formControl PodBrowser-adornedEnd PodBrowser-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="PodBrowser-input PodBrowser-input PodBrowser-input PodBrowser-inputFocused PodBrowser-inputAdornedEnd PodBrowser-inputAdornedEnd"
                id="provider-select"
                required=""
                spellcheck="false"
                type="url"
                value=""
              />
              <div
                class="PodBrowser-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="PodBrowser-root PodBrowser-root PodBrowser-clearIndicator"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="PodBrowser-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="PodBrowser-root PodBrowser-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="PodBrowser-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PodBrowser-root PodBrowser-notchedOutline"
              >
                <legend
                  class="PodBrowser-legendLabelled"
                >
                  <span>
                    Select ID provider *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          role="button"
          tabindex="0"
        >
          <button
            class="PodBrowser-button"
            data-testid="login-button"
            type="submit"
          >
            Log In
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
